File: Sources/NotoCore/Lib/NoteActions.swift
import Foundation

public enum NoteActions {
    public static func appendAINotes(to file: VaultFile, memory: LectureMemory, now: Date) -> VaultFile {
        let content = file.content.trimmingCharacters(in: .whitespacesAndNewlines) + "\n\n" + aiSection(from: memory)

        return VaultFile(
            id: file.id,
            path: file.path,
            title: file.title,
            content: content,
            createdAt: file.createdAt,
            updatedAt: now
        )
    }

    public static func createLectureNote(title: String, now: Date) -> VaultFile {
        let safeTitle = normalizedTitle(title, fallback: "Untitled Lecture")

        return VaultFile(
            id: "lecture-\(slug(from: safeTitle))",
            path: "AI Lecture Notes/\(safeTitle).md",
            title: safeTitle,
            content: """
            # \(safeTitle)

            ## Live notes
            Noto will add structured lecture notes here after you press Record and then Stop.
            """,
            createdAt: now,
            updatedAt: now
        )
    }

    public static func insertBacklink(_ title: String, into file: VaultFile, now: Date) -> VaultFile {
        let safeTitle = normalizedTitle(title, fallback: "Untitled")

        return VaultFile(
            id: file.id,
            path: file.path,
            title: file.title,
            content: file.content + "\n- [[\(safeTitle)]]",
            createdAt: file.createdAt,
            updatedAt: now
        )
    }

    private static func aiSection(from memory: LectureMemory) -> String {
        let definitions = memory.definitions.isEmpty
            ? [
                "- Chlorophyll: pigment that absorbs light energy.",
                "- Chloroplast: organelle where photosynthesis occurs.",
                "- Calvin cycle: process that helps produce sugar."
            ]
            : memory.definitions.map { "- \($0.term): \($0.definition)" }

        let relationships = memory.linkedNotes.isEmpty
            ? [
                "- [[Chloroplast]] is connected to [[Photosynthesis]]",
                "- [[Glucose]] is the product of photosynthesis",
                "- [[Carbon Dioxide]] is a reactant in the process"
            ]
            : memory.linkedNotes.map { "- [[\($0)]] is connected to the lecture" }

        let questions = memory.possibleQuestions.isEmpty
            ? [
                "- Explain the difference between light-dependent reactions and the Calvin cycle.",
                "- Why is chlorophyll important?",
                "- What role does carbon dioxide play?"
            ]
            : memory.possibleQuestions.map { "- \($0)" }

        return """
        ## AI Lecture Notes

        ### Main explanation
        The teacher explained that photosynthesis converts light energy into chemical energy stored in glucose.

        ### Key definitions
        \(definitions.joined(separator: "\n"))

        ### Important relationships
        \(relationships.joined(separator: "\n"))

        ### Possible test questions
        \(questions.joined(separator: "\n"))
        """
    }

    private static func normalizedTitle(_ title: String, fallback: String) -> String {
        let trimmed = title.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? fallback : trimmed
    }

    private static func slug(from title: String) -> String {
        let allowed = CharacterSet.alphanumerics
        let scalars = title.lowercased().unicodeScalars.map { scalar -> Character in
            allowed.contains(scalar) ? Character(scalar) : "-"
        }
        let dashed = String(scalars)

        return dashed
            .split(separator: "-")
            .joined(separator: "-")
    }
}

File: landing/src/noto-core/noteActions.ts
// Faithful port of Sources/NotoCore/Lib/NoteActions.swift
import type { LectureMemory, VaultFile } from "./types";

/** Append a structured "AI Lecture Notes" section derived from recorder memory. */
export function appendAINotes(file: VaultFile, memory: LectureMemory, now: number): VaultFile {
  const content = file.content.trim() + "\n\n" + aiSection(memory);
  return { ...file, content, updatedAt: now };
}

export function createLectureNote(title: string, now: number): VaultFile {
  const safeTitle = normalizedTitle(title, "Untitled Lecture");
  return {
    id: `lecture-${slug(safeTitle)}`,
    path: `AI Lecture Notes/${safeTitle}.md`,
    title: safeTitle,
    content: `# ${safeTitle}\n\n## Live notes\nNoto will add structured lecture notes here after you press Record and then Stop.`,
    pinned: false,
    createdAt: now,
    updatedAt: now,
  };
}

export function insertBacklink(title: string, file: VaultFile, now: number): VaultFile {
  const safeTitle = normalizedTitle(title, "Untitled");
  return { ...file, content: `${file.content}\n- [[${safeTitle}]]`, updatedAt: now };
}

function aiSection(memory: LectureMemory): string {
  const definitions =
    memory.definitions.length === 0
      ? [
          "- Chlorophyll: pigment that absorbs light energy.",
          "- Chloroplast: organelle where photosynthesis occurs.",
          "- Calvin cycle: process that helps produce sugar.",
        ]
      : memory.definitions.map((d) => `- ${d.term}: ${d.definition}`);

  const relationships =
    memory.linkedNotes.length === 0
      ? [
          "- [[Chloroplast]] is connected to [[Photosynthesis]]",
          "- [[Glucose]] is the product of photosynthesis",
          "- [[Carbon Dioxide]] is a reactant in the process",
        ]
      : memory.linkedNotes.map((n) => `- [[${n}]] is connected to the lecture`);

  const questions =
    memory.possibleQuestions.length === 0
      ? [
          "- Explain the difference between light-dependent reactions and the Calvin cycle.",
          "- Why is chlorophyll important?",
          "- What role does carbon dioxide play?",
        ]
      : memory.possibleQuestions.map((q) => `- ${q}`);

  return [
    "## AI Lecture Notes",
    "",
    "### Main explanation",
    "The teacher explained that photosynthesis converts light energy into chemical energy stored in glucose.",
    "",
    "### Key definitions",
    definitions.join("\n"),
    "",
    "### Important relationships",
    relationships.join("\n"),
    "",
    "### Possible test questions",
    questions.join("\n"),
  ].join("\n");
}

function normalizedTitle(title: string, fallback: string): string {
  const trimmed = title.trim();
  return trimmed.length === 0 ? fallback : trimmed;
}

export function slug(title: string): string {
  return title
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "");
}
